Fix attribute parser.

kindaro opened the pull request:
Hello.

I believe I found and fixed some bugs in attribute parser.

1. It won't handle a whitespace after attributes. (Because `sepBy` will eat the whitespace as a separator and try to parse whatever goes next as another attribute.)
2. It won't handle an unquoted value in the last attribute of a void element. (Because it will consider slash to be part of the value.)

I also made some non-critical changes.

1. Version bumped the library and its dependencies to 2.0.
2. Improved error reporting.
3. Aligned unquoted attribute value parser with HTML5 specification.

Please let me know in what way I should improve my patch for it to be merged.

kindaro committed:
Version bump to 2.0 and rename.

kindaro committed:
Improve error reporting.

kindaro committed:
Fix: whitespace after attributes breaks parsing.

kindaro committed:
Fix: constrain unquoted attribute value.

1. Treat as whitespace not only spacebar but also tabulation and carriage returns.
2. Consider slash be the end of attribute value. (Otherwise void tags won't parse.)
3. Consider unquoted attribute value unempty. (Otherwise we couldn't allow whitespace between
equals sign and unquoted attribute value.)

kindaro committed:
Fix: Wrong characters in unquoted attribute value.

According to HTML5 specification, unquoted attribute value cannot contain certain characters.

kindaro committed:
Add type signature for Test.Main.main function.

kindaro committed:
Add more tests on attribute value.

kindaro committed:
Fix: Exclude wrong test.

As (according to HTML5 specification) we can't have certain characters in unquoted attribute
value, we should remove them from our test case.

The spec: https://html.spec.whatwg.org/multipage/syntax.html#syntax-attributes

carymrobbins commented:
Great work! I'm pretty close to merging it, I just have a few questions.

What version of purescript did you use? The build is failing for me with 0.9.1 as well as with 0.10.3 (the latest). We need to get it to the latest compiler anyway, but I can do that once I get it to compile with whichever version you're using.

Also, I noticed that when I install dependencies with bower I now need to resolve a conflict with purescript-lists, so you'll want to pin the right version (usually the latest) in the resolutions section of bower.json.

See: http://stackoverflow.com/questions/16991128/how-to-resolve-bower-dependency-version-conflicts

Once this is all said and done I'll update the README with clear instructions on building from scratch, as I've had to figure it out again since it's been a while.

kindaro committed:
Fix purescript-lists version conflict.

kindaro committed:
Fix: whitespace after equals sign breaks parsing.

Whitespace after equals sign in an argument is not skipped correctly if the argument value is
unquoted. If we refactor a bit, the bug disappears.

kindaro committed:
Add a test case.

Add a test case for skipping whitespace around the equals sign in an argument with unquoted value.

kindaro commented:
(Aside: just noticed and solved another small bug in argument parser.)

I've been using 0.10.2. Just updated to 0.10.3 and it compiles with `pulp build` and passes all tests with `pulp test`. It does give 52 warnings though...

```
% git clone 'https://github.com/kindaro/purescript-html-parser' &&
> cd purescript-html-parser &&
> bower install &&
> pulp test
...
* Building project in /tmp/purescript-html-parser
...
* Build successful.
* Running tests...
...
* Tests OK.
```

As straightforward as can be...
